xdsl/frontend/python_code_check.py

```diff
@@ -141,3 +145,125 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
                                     inner_stmt.lineno, inner_stmt.col_offset,
                                     f"Cannot have an inner function '{inner_stmt.name}' inside the block '{stmt.name}'."
                                 )
+
+
+@dataclass
+class CheckAndInlineConstants:
+
+    @staticmethod
+    def run(stmts: List[ast.stmt]) -> None:
+        CheckAndInlineConstants.run_with_variables(stmts, set())
+
+    @staticmethod
+    def run_with_variables(stmts: List[ast.stmt],
+                           defined_variables: Set[str]) -> None:
+        for i, stmt in enumerate(stmts):
+            # This variable (`a = ...`) can be redefined as a constant, and so we have to
+            # keep track of these to raise an exception.
+            if isinstance(stmt, ast.Assign) and len(
+                    stmt.targets) == 1 and isinstance(stmt.targets[0],
+                                                      ast.Name):
+                defined_variables.add(stmt.targets[0].id)
+                continue
+
+            # Similarly, this case (`a: i32 = ...`) can also be redefined as a constant.
+            if isinstance(stmt, ast.AnnAssign) and isinstance(
+                    stmt.target,
+                    ast.Name) and not is_constant(stmt.annotation):
+                defined_variables.add(stmt.target.id)
+                continue
+
+            # This is a constant.
+            if isinstance(stmt, ast.AnnAssign) and is_constant(
+                    stmt.annotation):
+                if not isinstance(stmt.target, ast.Name):
+                    raise CodeGenerationException(
+                        stmt.lineno, stmt.col_offset,
+                        f"All constant expressions have to be assigned to 'ast.Name' nodes."
+                    )
+
+                name = stmt.target.id
+                try:
+                    value = eval(ast.unparse(stmt.value))
+                except Exception:
+                    # TODO: This error message can be improved by matching exact
+                    # exceptions returned by `eval` call.
+                    raise CodeGenerationException(
+                        stmt.lineno, stmt.col_offset,
+                        f"Non-constant expression cannot be assigned to constant variable '{name}' or cannot be evaluated."
```

Comment:
I would break that line in two (and other lines) so they fit in the 80 characters limit.